Zend/zend_execute.c

```diff
@@ -1874,13 +1888,23 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EG(current_execute_data)->opline->op1.var, EG(current_execute_data));
+			} else if (!zend_fetch_dimension_address_read_valid_container(container, support_strings)) {
+				zend_error(E_WARNING, "Variable of type %s does not accept array offsets", zend_get_type_by_const(Z_TYPE_P(container)));
+			}
 		}
 		if (/*dim_type == IS_CV &&*/ UNEXPECTED(Z_TYPE_P(dim) == IS_UNDEF)) {
 			zval_undefined_cv(EG(current_execute_data)->opline->op2.var, EG(current_execute_data));
 		}
-		ZVAL_NULL(result);
+
+		/* return error if we're continuing looking up.  This is OK since after FETCH_DIM_R there will be a FREE op */
+		if ((EG(current_execute_data)->opline + 1)->opcode == ZEND_FETCH_DIM_R) {
+			ZVAL_ERROR(result);
```

Comment:
I afraid, this hack may make a lot of unexpected side effects.

$ php -r '$a = [null]; $c = null; var_dump($a[0][0] + $c[0]);'
int(0)

$ sapi/cli/php -r '$a = [null]; $c = null; var_dump($a[0][0] + $c[0]);'
PHP Warning:  Variable of type null does not accept array offsets in Command line code on line 1

Warning: Variable of type null does not accept array offsets in Command line code on line 1
PHP Warning:  Variable of type null does not accept array offsets in Command line code on line 1

Warning: Variable of type null does not accept array offsets in Command line code on line 1
PHP Fatal error:  Uncaught Error: Unsupported operand types in Command line code:1
Stack trace:
#0 {main}
  thrown in Command line code on line 1

Fatal error: Uncaught Error: Unsupported operand types in Command line code:1
Stack trace:
#0 {main}
  thrown in Command line code on line 1

@nikic ^